### PR TITLE
 Introspector in swarm

### PR DESCRIPTION
**Pull request by aarshkshah1992**
For https://github.com/libp2p/go-libp2p/issues/775

Changes for swarm to be able to register as a data provider & then provide the data/metrics when it is asked to do so.

### CODE REVIEW

**Comment by aarshkshah1992** [comment]
### **TODO**
- [ ] Release go-libp2p-core & configure the version here
- [ ] Build the provider for connections & stream
- [ ] Tests for the providers

File: swarm.go [hunk 1]
```diff
@@ -9,6 +9,7 @@ import (
 	"sync/atomic"
 	"time"
 
+	coreit "github.com/libp2p/go-libp2p-core/introspection"
```
Comment (raulk) [inline swarm.go]: Shorten `introspection` to `introspect`, and drop the import alias.

File: swarm.go [hunk 2]
```diff
@@ -115,6 +116,13 @@ func NewSwarm(ctx context.Context, local peer.ID, peers peerstore.Peerstore, bwc
 	s.proc = goprocessctx.WithContextAndTeardown(ctx, s.teardown)
 	s.ctx = goprocessctx.OnClosingContext(s.proc)
 
+	// TODO register provider for introspection
+	if introspector != nil {
```
Comment (raulk) [inline swarm.go]: Would love to see the actual provider functions, after my comments on https://github.com/libp2p/go-libp2p-core/pull/107 have been integrated.

**Comment by aarshkshah1992** [comment]
@raulk  Here are the fields that still need some discussion before we implement sourcing for them:

### Stream

1.  **UserProvidedTags**
    What does this mean ?
2.

File: conn_introspect.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+func (ci *connIntrospector) id() string {
+	return ci.c.id
+}
+
+func (ci *connIntrospector) remotePeerId() string {
+	return ci.c.RemotePeer().String()
+}
+
+func (ci *connIntrospector) endPoints() *introspectpb.EndpointPair {
+	return &introspectpb.EndpointPair{SrcMultiaddr: ci.c.LocalMultiaddr().String(), DstMultiaddr: ci.c.RemoteMultiaddr().String()}
```
Comment (raulk) [inline conn_introspect.go]: nit: multi-line init is more readable in this case.

File: conn_introspect.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+		return introspectpb.Role_RESPONDER
+	} else {
+		return introspectpb.Role_INITIATOR
+	}
+}
+
+// TODO Number of packets & instantaneous bandwidth ?
+// Should we have a separate "flow-metre" for a connection to prepare for a message oriented world ?
+func (ci *connIntrospector) traffic() *introspectpb.Traffic {
+	if ci.s.bwc != nil {
```
Comment (raulk) [inline conn_introspect.go]: In these cases, prefer short-circuiting to reduce one level of nesting/indentation.

```go
if ci.s.bwc == nil {
  return nil
}
<rest>
```

File: introspect.go [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+import (
+	"github.com/libp2p/go-libp2p-core/introspect"
+	introspectpb "github.com/libp2p/go-libp2p-core/introspect/pb"
+	"github.com/pkg/errors"
+)
+
+// IntrospectTraffic introspects & returns the overall traffic for this peer
+func (s *Swarm) IntrospectTraffic() (*introspectpb.Traffic, error) {
+	if s.bwc != nil {
```
Comment (raulk) [inline introspect.go]: Reverse the condition, and outdent the logic outside the if.

File: introspect.go [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+		t := &introspectpb.Traffic{}
+		metrics := s.bwc.GetBandwidthTotals()
+
+		t.TrafficIn = &introspectpb.DataGauge{CumBytes: uint64(metrics.TotalIn), InstBw: uint64(metrics.RateIn)}
```
Comment (raulk) [inline introspect.go]: nit: multi-line init is more readable.

File: introspect.go [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+	return nil, nil
+}
+
+// IntrospectConnections introspects & returns the swarm connections
+func (swarm *Swarm) IntrospectConnections(query introspect.ConnectionQueryParams) ([]*introspectpb.Connection, error) {
+	swarm.conns.RLock()
+	defer swarm.conns.RUnlock()
+
+	containsId := func(ids []introspect.ConnectionID, id string) bool {
```
Comment (raulk) [inline introspect.go]: More efficient to create a `map[introspect.ConnectionID]struct{}`, and make the predicate test membership. This is `O(m*n)` worst case scenario; map-based solution would be `O(m)` at the expense of a little extra memory.

File: introspect_test.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
+package swarm_test
+
+import (
+	"context"
```
Comment (raulk) [inline introspect_test.go]: Regroup imports.

File: swarm.go [hunk 9]
```diff
@@ -209,11 +222,13 @@ func (s *Swarm) addConn(tc transport.CapableConn, dir network.Direction) (*Conn,
 	}
 
 	// Wrap and register the connection.
+	id := uuid.NewV4()
```
Comment (raulk) [inline swarm.go]: An UUID is pretty inefficient. Requires string comparison. Why not an int32 atomic counter at the swarm level, which you increment with every new connection?

**Comment by aarshkshah1992** [comment]
@raulk I'm happy to refactor this PR along with implementing the stream/connection scoped metering if you like.

**Pull request merged**